mitiq/cdr/clifford_training_data.py

```diff
@@ -171,8 +169,7 @@ def _select(
                       method_select = 'gaussian'.
         random_state: Random state for sampling.
     """
-    if random_state is None:
-        random_state = np.random  # type: ignore
```

Comment:
The reason we used `random_state = np.random` instead of a new instance  `random_state = np.RandomState()` is that the user can set a global seed  `np.random.seed(123)` with a single instruction e.g. at the beginning of a notebook and it should propagate to all Mitiq code.
Cirq uses a similar [approach](https://github.com/quantumlib/Cirq/blob/f2cd706ecd308b8088f7452e02e6f24090087e86/cirq-core/cirq/value/random_state.py#L40).

Actually I am not sure if this is good or bad practice. Just wanted to explain the motivation.

Reply:
Thanks for adding some background here (and the link to Cirq), it's super helpful! I agree we should definitely tread lightly around changing RandomState related things, but I'm not sure that [`np.random.seed`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html) works in that way. From my understanding, this function only effects the _next_ call of a numpy random function, but not subsequent ones. So if a user sets a seed at the top of a file running Mitiq, I would not expect completely reproducible outputs.
```py
>>> import numpy as np
>>> np.random.seed(42)
>>> np.random.randint(500)
102
>>> np.random.seed(42)
>>> np.random.randint(500)
102
>>> np.random.randint(500)
435
```

Reply:
The seed actually affects subsequent functions (the internal random states evolves in a reproducible way):

```python
>>> import numpy as np

>>> np.random.seed(42)
>>> np.random.randint(500)
102
>>> np.random.randint(500)
435
>>> np.random.seed(42)
>>> np.random.randint(500)
102
>>> np.random.randint(500)
435
```

So, everything which depends on `np.random` becomes deterministically reproducible.

Reply:
I tried `random_state: Union[np.random.RandomState, np.random.Generator] = np.random.default_rng(),` and all tests pass. Can remove the None checks and most Optionals too, eg:
```
def _select(
    non_clifford_ops: Sequence[cirq.ops.Operation],
    fraction_non_clifford: float,
    method: str = "uniform",
    sigma: Optional[float] = 1.0,
    random_state: Union[np.random.RandomState, np.random.Generator] = np.random.default_rng(),
) -> List[int]:
...
    num_non_cliff = len(non_clifford_ops)
    num_to_replace = int(round(fraction_non_clifford * num_non_cliff))
```
Happy to push if we like it and you're low on bandwidth @natestemen.

Reply:
@natestemen, as discussed made a branch to test out these changes. They're [here](https://github.com/unitaryfund/mitiq/tree/ar-clean-types) when you get a chance to take a look!

Reply:
From my testing the `np.random.default_rng` function does not respect if the user has set a seed using `np.random.seed`. So I'll revert my changes here back to what we had previously.